lualibs/go/atc-router/api.go

```diff
@@ -0,0 +1,106 @@
+package atcrouter
+
+import (
+	"fmt"
+
+	"github.com/gofrs/uuid"
+	router "github.com/kong/go-atc-router"
+	lua "github.com/yuin/gopher-lua"
+)
+
+// lua: schema = new_schema()
+func newSchema(l *lua.LState) int {
+	schemaUD := l.NewUserData()
+	schemaUD.Value = router.NewSchema()
+	schemaUD.Metatable = l.Get(lua.EnvironIndex)
+
+	l.Push(schemaUD)
+	return 1
+}
+
+func raiseUDArgError(l *lua.LState, n int, expected string, got *lua.LUserData) {
+	l.ArgError(n, fmt.Sprintf("%v expected, got %T", expected, got.Value))
+}
+
+// lua: schema:add_field(fldname, type)
+func schemaAddField(l *lua.LState) int {
+	const (
+		schemaArgNum    = 1
+		fieldNameArgNum = 2
+		fieldTypeArgNum = 3
+	)
+
+	schemaUD := l.CheckUserData(schemaArgNum)
+	schema, ok := schemaUD.Value.(*router.Schema)
+	if !ok {
+		raiseUDArgError(l, schemaArgNum, "schema", schemaUD)
```

Comment:
Missing a bunch of early returns in error handling.

Reply:
the error raising functions don't return.   internally, the `LState.raiseError()` function ends in a `panic()` which is caught at the `PCall()` function (the one that calls Go functions from Lua).

Reply:
Why so? For this particular example, this code would panic in the event it's not okay? Why deviate from best practices, even if there is a recovery handler elsewhere?

Reply:
that's how gopher-lua is written.

in Lua, it's considered "best practice" to raise an error on issues that are most likely to be programming errors, like required arguments missing or of the wrong type.  for other more "runtime related" errors, the (falsy, message) idiom is preferred.

Reply:
in this specific code, an error would be raised if the Lua code tries to add a field to something that is not a schema object, or if the type argument isn't one of the known options.

this is the same behaviour as the Lua FFI wrapping in https://github.com/Kong/atc-router/blob/main/lib/resty/router/schema.lua#L28